source/Timer.cpp

```diff
@@ -0,0 +1,346 @@
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "Timer.h"
+
+#include "DataNode.h"
+#include "DataWriter.h"
+#include "GameData.h"
+#include "Logger.h"
+#include "Planet.h"
+#include "PlayerInfo.h"
+#include "Random.h"
+#include "Ship.h"
+#include "StellarObject.h"
+#include "UI.h"
+
+using namespace std;
+
+namespace {
+	string TriggerToText(Timer::TimerTrigger trigger)
+	{
+		switch(trigger)
+		{
+			case Timer::TimerTrigger::TIMEUP:
+				return "on timeup";
+			case Timer::TimerTrigger::RESET:
+				return "on reset";
+			default:
+				return "unknown trigger";
+		}
+	}
+}
+
+
+
+Timer::Timer(const DataNode &node)
+{
+	Load(node);
+}
+
+
+
+void Timer::Load(const DataNode &node)
+{
+	for(const DataNode &child : node)
+	{
+		if(child.Token(0) == "time" && child.Size() > 1)
+		{
+			waitTime = child.Value(1);
+			if(child.Size() > 2)
+				randomWaitTime = child.Value(2);
+		}
+		else if(child.Token(0) == "idle")
+		{
+			requireIdle = true;
+			// We square the max speed value here, so it can be conveniently
+			// compared to the flagship's squared velocity length below.
+			if(child.Size() > 1)
+				idleMaxSpeed = child.Value(1) * child.Value(1);
+		}
+		else if(child.Token(0) == "peaceful")
+			requirePeaceful = true;
+		else if(child.Token(0) == "optional")
+			optional = true;
+		else if(child.Token(0) == "system" && child.Size() > 1)
+			system = GameData::Systems().Get(child.Token(1));
+		else if(child.Token(0) == "system")
+			systems.Load(child);
+		else if(child.Token(0) == "proximity" && child.Size() > 1)
+		{
+			proximityCenter = GameData::Planets().Get(child.Token(1));
+		}
+		else if(child.Token(0) == "proximity")
+		{
+			proximityCenters.Load(child);
+		}
+		else if(child.Token(0) == "proximity settings" && child.Size() > 1)
+		{
+			proximity = child.Value(1);
+			if(child.Size() > 2)
+			{
+				if(child.Token(2) == "close")
+					closeTo = true;
+				else if(child.Token(2) == "far")
+					closeTo = false;
+			}
+		}
+		else if(child.Token(0) == "uncloaked")
+			requireUncloaked = true;
+		else if(child.Token(0) == "reset" && child.Size() > 1)
+		{
+			const string &value = child.Token(1);
+			if(value == "leave system")
+				resetCondition = ResetCondition::LEAVE_SYSTEM;
+			else if(value == "leave zone")
+				resetCondition = ResetCondition::LEAVE_ZONE;
+			else if(value == "none")
+				resetCondition = ResetCondition::NONE;
+			else if(value == "pause")
+				resetCondition = ResetCondition::PAUSE;
+			else
+				child.PrintTrace("Skipping unrecognized attribute:");
```

Comment:
wouldnt it be possible to have multiple of these possible?

Reply:
Hhhhypothetically, I believe so? I think the only one that could coexist with others in practice is `PAUSE`—leaving the system and leaving the zone are basically just different degrees of moving too far away, but pausing the count (ie, moving if it requires idle, firing if it requires peaceful) _could_ be tested for separately.

Does that seem like it's worth refactoring to add the functionality?

Reply:
Nah